File: Pinball App 2/Pinball App 2/library/LibraryResourcePathSupport.swift
import Foundation

nonisolated let librarySupportedPlayfieldOriginalExtensions = ["webp", "jpg", "jpeg", "png"]
nonisolated let libraryMissingArtworkPath = "/pinball/images/playfields/fallback-image-not-available_2048.webp"
nonisolated let libraryPinProfHosts: Set<String> = [
    "pillyliu.com",
    "www.pillyliu.com",
    "pinprof.com",
    "www.pinprof.com"
]
nonisolated let libraryBundledOnlyAppGroupIDs: Set<String> = [
    "G900001"
]

enum LibraryLivePlayfieldKind: String {
    case pillyliu
    case opdb
    case external
    case missing
}

struct LibraryLivePlayfieldStatus: Equatable {
    let effectiveKind: LibraryLivePlayfieldKind
    let effectiveURL: URL?
}

actor LibraryLivePlayfieldStatusStore {
    static let shared = LibraryLivePlayfieldStatusStore()

    func status(for practiceIdentity: String?) async -> LibraryLivePlayfieldStatus? {
        guard let practiceIdentity = practiceIdentity?.trimmingCharacters(in: .whitespacesAndNewlines),
              !practiceIdentity.isEmpty else {
            return nil
        }

        var components = URLComponents(string: "https://pillyliu.com/pinprof-admin/api.php")
        components?.queryItems = [
            URLQueryItem(name: "route", value: "public/playfield-status/\(practiceIdentity)")
        ]
        guard let url = components?.url else { return nil }

        var request = URLRequest(url: url)
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.timeoutInterval = 15
        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode),
                  let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let rawKind = json["effectiveKind"] as? String,
                  let kind = LibraryLivePlayfieldKind(rawValue: rawKind) else {
                return nil
            }

            let effectiveURL = libraryResolveURL(pathOrURL: (json["effectiveUrl"] as? String) ?? "")
            return LibraryLivePlayfieldStatus(effectiveKind: kind, effectiveURL: effectiveURL)
        } catch {
            return nil
        }
    }
}

nonisolated func libraryResolveURL(pathOrURL: String) -> URL? {
    if let direct = URL(string: pathOrURL), direct.scheme != nil {
        return direct
    }

    if pathOrURL.hasPrefix("/") {
        return URL(string: "https://pillyliu.com\(pathOrURL)")
    }

    return URL(string: "https://pillyliu.com/\(pathOrURL)")
}

nonisolated func libraryIsPinProfHost(_ host: String?) -> Bool {
    guard let host = host?.lowercased() else { return false }
    return libraryPinProfHosts.contains(host)
}

nonisolated func libraryIsPinProfPlayfieldURL(_ url: URL?) -> Bool {
    guard let url,
          libraryIsPinProfHost(url.host) else {
        return false
    }
    return url.path.hasPrefix("/pinball/images/playfields/")
}

nonisolated func normalizeLibraryCachePath(_ pathOrURL: String?) -> String? {
    guard let raw = pathOrURL?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
        return nil
    }
    func normalizePlayfieldPublishedPath(_ value: String) -> String {
        value.replacingOccurrences(
            of: #"(/pinball/images/playfields/.+?)(?:_(700|1400))?\.[A-Za-z0-9]+$"#,
            with: "$1.webp",
            options: .regularExpression
        )
    }
    if let url = URL(string: raw), let host = url.host?.lowercased(), host == "pillyliu.com" {
        return url.path.contains("/pinball/images/playfields/") ? normalizePlayfieldPublishedPath(url.path) : url.path
    }
    if raw.hasPrefix("/") {
        return raw.contains("/pinball/images/playfields/") ? normalizePlayfieldPublishedPath(raw) : raw
    }
    let normalized = "/" + raw
    return normalized.contains("/pinball/images/playfields/") ? normalizePlayfieldPublishedPath(normalized) : normalized
}

nonisolated func normalizeLibraryPlayfieldLocalPath(_ pathOrURL: String?) -> String? {
    guard let raw = pathOrURL?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
        return nil
    }
    if raw.localizedCaseInsensitiveContains("/pinball/images/playfields/") {
        if raw.lowercased().hasSuffix("_700.webp") { return raw }
        if raw.lowercased().hasSuffix("_1400.webp") {
            return raw.replacingOccurrences(of: "_1400.webp", with: "_700.webp", options: [.caseInsensitive])
        }
        if let dot = raw.lastIndex(of: ".") {
            return String(raw[..<dot]) + "_700.webp"
        }
    }
    return raw
}

nonisolated func libraryMissingArtworkURL() -> URL? {
    libraryResolveURL(pathOrURL: libraryMissingArtworkPath)
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibraryResourcePathSupport.kt
package com.pillyliu.pinprofandroid.library

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URLEncoder
import java.net.URL

internal const val libraryMissingArtworkPath = "/pinball/images/playfields/fallback-image-not-available_2048.webp"

private val pinProfHosts = setOf(
    "pillyliu.com",
    "www.pillyliu.com",
    "pinprof.com",
    "www.pinprof.com",
)

internal enum class LivePlayfieldKind {
    PILLYLIU,
    OPDB,
    EXTERNAL,
    MISSING,
}

internal data class LivePlayfieldStatus(
    val effectiveKind: LivePlayfieldKind,
    val effectiveUrl: String?,
)

internal suspend fun loadLivePlayfieldStatus(practiceIdentity: String?): LivePlayfieldStatus? = withContext(Dispatchers.IO) {
    val normalizedPracticeIdentity = practiceIdentity?.trim()?.takeIf { it.isNotEmpty() } ?: return@withContext null
    val route = URLEncoder.encode("public/playfield-status/$normalizedPracticeIdentity", Charsets.UTF_8.name())
    val requestUrl = "https://pillyliu.com/pinprof-admin/api.php?route=$route"
    val connection = (URL(requestUrl).openConnection() as HttpURLConnection).apply {
        connectTimeout = 15_000
        readTimeout = 15_000
        requestMethod = "GET"
        setRequestProperty("Cache-Control", "no-cache")
    }
    try {
        val code = connection.responseCode
        if (code !in 200..299) return@withContext null
        val payload = connection.inputStream.bufferedReader().use { it.readText() }
        val json = JSONObject(payload)
        val kind = when (json.optString("effectiveKind").lowercase()) {
            "pillyliu" -> LivePlayfieldKind.PILLYLIU
            "opdb" -> LivePlayfieldKind.OPDB
            "external" -> LivePlayfieldKind.EXTERNAL
            "missing" -> LivePlayfieldKind.MISSING
            else -> return@withContext null
        }
        LivePlayfieldStatus(
            effectiveKind = kind,
            effectiveUrl = resolveLibraryUrl(json.optString("effectiveUrl").ifBlank { null }),
        )
    } catch (_: Throwable) {
        null
    } finally {
        connection.disconnect()
    }
}

internal fun resolveLibraryUrl(pathOrUrl: String?): String? {
    val normalized = normalizedOptionalString(pathOrUrl) ?: return null
    if (normalized.startsWith("http://") || normalized.startsWith("https://")) return normalized
    return if (normalized.startsWith("/")) "https://pillyliu.com$normalized" else "https://pillyliu.com/$normalized"
}

internal fun isPinProfHost(host: String?): Boolean =
    host?.lowercase()?.let(pinProfHosts::contains) == true

internal fun isPinProfPlayfieldUrl(url: String?): Boolean {
    val resolved = resolveLibraryUrl(url) ?: return false
    return runCatching {
        val parsed = URL(resolved)
        isPinProfHost(parsed.host) && parsed.path.startsWith("/pinball/images/playfields/")
    }.getOrDefault(false)
}

internal fun isPinProfRulesheetUrl(url: String?): Boolean {
    val resolved = resolveLibraryUrl(url) ?: return false
    return runCatching {
        val parsed = URL(resolved)
        isPinProfHost(parsed.host) && parsed.path.startsWith("/pinball/rulesheets/")
    }.getOrDefault(false)
}

internal fun normalizeLibraryPlayfieldLocalPath(path: String?): String? {
    val raw = normalizedOptionalString(path) ?: return null
    val target = when {
        raw.endsWith("_700.webp", ignoreCase = true) -> raw
        raw.endsWith("_1400.webp", ignoreCase = true) -> raw.replace(Regex("_1400\\.webp$", RegexOption.IGNORE_CASE), "_700.webp")
        raw.contains("/pinball/images/playfields/") -> raw.replace(Regex("\\.[A-Za-z0-9]+$"), "_700.webp")
        else -> raw
    }
    return target
}

internal fun normalizeLibraryCachePath(path: String?): String? {
    val raw = normalizedOptionalString(path) ?: return null
    fun normalizePlayfieldPublishedPath(value: String): String =
        value.replace(Regex("(/pinball/images/playfields/.+?)(?:_(700|1400))?\\.[A-Za-z0-9]+$", RegexOption.IGNORE_CASE), "$1.webp")
    if (raw.startsWith("/")) {
        return if (raw.contains("/pinball/images/playfields/")) normalizePlayfieldPublishedPath(raw) else raw
    }
    if (raw.startsWith("http://") || raw.startsWith("https://")) {
        return try {
            val uri = java.net.URI(raw)
            if (uri.host?.equals("pillyliu.com", ignoreCase = true) == true) {
                uri.path?.takeIf { it.isNotBlank() }?.let {
                    if (it.contains("/pinball/images/playfields/")) normalizePlayfieldPublishedPath(it) else it
                } ?: raw
            } else {
                raw
            }
        } catch (_: Exception) {
            raw
        }
    }
    val normalized = "/$raw"
    return if (normalized.contains("/pinball/images/playfields/")) normalizePlayfieldPublishedPath(normalized) else normalized
}

internal fun PinballGame.resolve(pathOrUrl: String?): String? =
    resolveLibraryUrl(pathOrUrl)

internal fun hostedRenderedRulesheetPageUrl(slug: String, source: String? = null): String {
    val encodedSlug = URLEncoder.encode(slug, Charsets.UTF_8.name())
    val encodedSource = source?.trim()?.takeIf { it.isNotEmpty() }
        ?.let { URLEncoder.encode(it, Charsets.UTF_8.name()) }
    return if (encodedSource != null) {
        "https://pillyliu.com/rules/$encodedSlug?source=$encodedSource"
    } else {
        "https://pillyliu.com/rules/$encodedSlug"
    }
}
